src/backend/distributed/commands/statistics.c

```diff
@@ -0,0 +1,243 @@
+/*-------------------------------------------------------------------------
+ *
+ * statistics.c
+ *    Commands for STATISTICS statements.
+ *
+ *    We currently support replicating statistics definitions on the
+ *    coordinator in all the worker nodes in the form of
+ *
+ *    CREATE STATISTICS ... queries.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/genam.h"
+#include "access/htup_details.h"
+#include "access/xact.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_namespace.h"
+#include "catalog/pg_statistic_ext.h"
+#include "catalog/pg_type.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/commands.h"
+#include "distributed/deparse_shard_query.h"
+#include "distributed/deparser.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/multi_executor.h"
+#include "distributed/namespace_utils.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/resource_lock.h"
+#include "distributed/worker_transaction.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/ruleutils.h"
+#include "utils/syscache.h"
+
+static List * GetExplicitStatisticsIdList(Oid relationId);
+static void EnsureSequentialModeForStatisticsDDL(void);
+
+
+/*
+ * PreprocessCreateStatisticsStmt is called during the planning phase for
+ * CREATE STATISTICS.
+ */
+List *
+PreprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	CreateStatsStmt *stmt = castNode(CreateStatsStmt, node);
+
+	DDLJob *ddlJob = palloc0(sizeof(DDLJob));
+	RangeVar *relation = (RangeVar *) linitial(stmt->relations);
+	Oid relationId = RangeVarGetRelid(relation, ShareUpdateExclusiveLock, false);
+
+	if (!IsCitusTable(relationId))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	QualifyTreeNode((Node *) stmt);
+
+	char *ddlCommand = DeparseTreeNode((Node *) stmt);
+
+	EnsureSequentialModeForStatisticsDDL();
+
+	ddlJob->targetRelationId = RangeVarGetRelid(relation, ShareUpdateExclusiveLock,
+												false);
+	ddlJob->concurrentIndexCmd = false;
+	ddlJob->startNewTransaction = false;
+	ddlJob->commandString = ddlCommand;
+	ddlJob->taskList = DDLTaskList(relationId, ddlCommand);
+
+	List *ddlJobs = list_make1(ddlJob);
+
+	return ddlJobs;
+}
+
+
+/*
+ * PostprocessCreateStatisticsStmt is called after a CREATE STATISTICS command has
+ * been executed by standard process utility.
+ */
+List *
+PostprocessCreateStatisticsStmt(Node *node, const char *queryString)
+{
+	CreateStatsStmt *stmt = castNode(CreateStatsStmt, node);
+	Assert(stmt->type == T_CreateStatsStmt);
+
+	if (!IsCoordinator() || !ShouldPropagate())
+	{
```

Comment:
This does not cover the case where the shard is on the coordinator.

Reply:
We could maybe repeat the IsCitusTable check, or set the EnableDDLPropagation GUC in worker_apply_ddl_command.